src/ERC4626StreamHub.sol

```diff
@@ -0,0 +1,206 @@
+        vault = _vault;
+        IERC20(vault.asset()).safeApprove(address(vault), type(uint256).max);
+    }
+
+    /**
+     * @dev Opens a yield stream for a specific receiver with a given number of shares.
+     * @param _receiver The address of the receiver.
+     * @param _shares The number of shares to allocate for the yield stream.
+     */
+    function openYieldStream(address _receiver, uint256 _shares) public {
+        _checkZeroAddress(_receiver);
+        _checkReceiver(_receiver);
+        _checkZeroShares(_shares);
+
+        vault.safeTransferFrom(msg.sender, address(this), _shares);
+
+        uint256 principal = _convertToAssets(_shares);
+
+        // balanceOf[msg.sender] -= _shares;
+        receiverShares[_receiver] += _shares;
+        receiverTotalPrincipal[_receiver] += principal;
+        receiverPrincipal[_receiver][msg.sender] += principal;
+
+        emit OpenYieldStream(msg.sender, _receiver, _shares, principal);
+    }
+
+    // TODO: do we need batch functions? same could be achieved using multicall
+    /**
+     * @dev Opens yield streams for multiple receivers with corresponding shares.
+     * @param _receivers An array of receiver addresses.
+     * @param _shares An array of share amounts corresponding to each receiver.
+     */
+    function openYieldStreamBatch(
+        address[] calldata _receivers,
+        uint256[] calldata _shares
+    ) external {
+        if (_receivers.length != _shares.length)
+            revert InputParamsLengthMismatch();
+
+        for (uint256 i = 0; i < _receivers.length; i++) {
+            openYieldStream(_receivers[i], _shares[i]);
+        }
+    }
+
+    /**
+     * @dev Closes a yield stream for a specific receiver.
+     * If there is any yield to claim for the stream, it will remain unclaimed until the receiver calls `claimYield` function.
+     * @param _receiver The address of the receiver.
+     */
+    function closeYieldStream(address _receiver) public {
+        uint256 principal = receiverPrincipal[_receiver][msg.sender];
+
+        if (principal == 0) revert StreamDoesNotExist();
+
+        // asset amount of equivalent shares
+        uint256 ask = _convertToShares(principal);
+        uint256 totalPrincipal = receiverTotalPrincipal[_receiver];
+        // the maximum amount of shares that can be attributed to the sender
+        uint256 have = receiverShares[_receiver].mulDivDown(
+            principal,
+            totalPrincipal
+        );
+
+        // if there was a loss, withdraw the percentage of the shares
+        // equivalent to the sender share of the total principal
+        uint256 shares = ask > have ? have : ask;
+
+        // update state and transfer
+        receiverPrincipal[_receiver][msg.sender] = 0;
+        receiverTotalPrincipal[_receiver] -= totalPrincipal;
+        receiverShares[_receiver] -= shares;
+
+        vault.safeTransfer(msg.sender, shares);
+
+        emit CloseYieldStream(msg.sender, _receiver, shares);
+    }
+
```

Comment:
also you removed previewClose, but why. its kind of useful. i dont get why you didnt just copy what what i did and then add the improvements. would have been less work 😉 also easier to follow since now all variable names changed

Reply:
I agree, you're right about the previewClose function, it's added back. And I should've stuck closer to what you did with the code, now I feel like I was trying to reinvent the wheel :)